procmonAnalyze.js

```diff
@@ -413,33 +472,73 @@ function drawBackground() {
 
   gState.timelineIndicators = [];
   timeline.textContent = "";
-  let printSeconds = timelineScale == 1;
-  for (let i = 0; i < Math.floor(totalTime / timelineScale); i++) {
-    let offset = i * timelineScale;
+  let offset;
+  for (let i = 0; i < Math.ceil(totalTime / timelineScale); i++) {
+    offset = i * timelineScale - initialHeight;
     let offsetPx = (offset + rendererTranslate) * rendererScale;
-    if (offsetPx < -VIEWPORT_BUFFER || offsetPx > canvas.height + VIEWPORT_BUFFER) {
-      continue;
-    }
 
     let div = document.createElement("div");
     div.style.position = "fixed";
-    div.style.left = `${canvas.width - 64}px`;
+    div.style.left = `${canvas.width - 40}px`;
     div.style.top = `${offsetPx}px`;
-    div.textContent = timelineScale == 1 ? `${i}s` : `${Math.round(i * timelineScale * 1000)}ms`;
+    div.textContent = timelineScale == 1 ? `${i + Math.floor(minTimeRelative)}s` : `${Math.round((i + Math.ceil(minTimeRelative)) * timelineScale * 1000)}ms`;
 
     timeline.appendChild(div);
     gState.timelineIndicators.push({div, offset});
   }
+
+  // Add the values of the range markers
+  if (isRefocused) {
+    let upperRangeDiv = document.createElement("div");
+    let offsetUpperPx = (rendererTranslate) * rendererScale;
+    upperRangeDiv.style.position = "fixed";
+    upperRangeDiv.style.left = `${canvas.width - 80}px`;
+    upperRangeDiv.style.top = `${offsetUpperPx}px`;
+    upperRangeDiv.textContent = timelineScale == 1 ? `${minTimeRelative.toFixed(2)}s` : `${(minTimeRelative * timelineScale * 1000).toFixed(2)}ms`;
+    timeline.appendChild(upperRangeDiv);
+    gState.timelineIndicators.push({upperRangeDiv, offset:0});
+
+    let lowerRangeDiv = document.createElement("div");
+    let offsetLowerPx = (offset + rendererTranslate) * rendererScale;
+    lowerRangeDiv.style.position = "fixed";
+    lowerRangeDiv.style.left = `${canvas.width - 80}px`;
+    lowerRangeDiv.style.top = `${canvas.height}px`;
+    lowerRangeDiv.textContent = timelineScale == 1 ? `${maxTimeRelative.toFixed(2)}s` : `${(maxTimeRelative * timelineScale * 1000).toFixed(2)}ms`;
+    timeline.appendChild(lowerRangeDiv);
+    gState.timelineIndicators.push({lowerRangeDiv, offset});    
+  }
+}
+
+
+function drawRangeMarkers() {
+  if (markerCtx) {
+    markerCtx.clearRect(0,0,markersCanvas.width, markersCanvas.height);
+  } else {
+    markerCtx = markersCanvas.getContext('2d');
+  }
+
+  var cw = markersCanvas.width;
+  var ch = markersCanvas.height;
+  reOffset();
+  topRange = makeRangeControl(0,0,cw,ch, true);
+  bottomRange = makeRangeControl(0,0,cw,ch, false)
+  drawRangeControl(topRange, true);
+  drawRangeControl(bottomRange, false);
+
+
+  markerCtx.onmousedown=(function(e){handleMouseDown(e);});
+  markerCtx.onmousemove=(function(e){handleMouseMove(e);});
+  markerCtx.onmouseup=(function(e){handleMouseUpOut(e);});
+  markerCtx.onmouseout=(function(e){handleMouseUpOut(e);});
```

Comment:
I think this is a bit problematic... registering these handlers here conflicts with the registrations on the `document` below. Can we just integrate the marker handling logic into the existing mouse handlers? I think the way it's set up here also causes problems because if you mouse out of the marker canvas, it will stop adjusting. You should be able to click and drag anywhere and have it move the marker just according to your mouse's Y position.

Reply:
Also, as a general note: I would keep the registration of event handlers outside of any `draw...` functions. It's a bit of a misleading name, and you end up setting the event handlers more than once, which is less than ideal.